fix issues #10442: no or incomplete RTInfo

rainers opened the pull request:
build TypeInfo in semantic, not during code-gen. This fixes RTInfo generation used by the precise GC implementation.

http://d.puremagic.com/issues/show_bug.cgi?id=10786

happens if the TypeInfo is referred to during the semantic3 pass, but is added to a module that has already completed the semantic3 step. This is fixable using the new deferred-semantic3 list. But the fixed version often triggers the next bug:

http://d.puremagic.com/issues/show_bug.cgi?id=10442

The problem is that the TypeInfo records are generated only from the glue or backend layer (e2ir.c or toobj.c) by calling Type::getTypeInfo(NULL), which outputs the TypeInfo without further analysis to the object file.
This ignores the RTInfo definition and just writes 0/1 depending on whether there are pointer fields in the struct/class or not.

Assuming that it is a bug to call Type::getTypeInfo(NULL) after semantic3 is finished, this patch forces the creation of the TypeInfo records during semantic for those expressions that later need them in the glue layer. A
problem with that approach is that it adds the TypeInfo of imported code into the object file (as COMDAT) if it is used in CTFE or type inference (similar to template functions). This makes http://d.puremagic.com/issues/show_bug.cgi?id=10831 show up more often.

Another problem is that the type of expressions is often changed later by implicite/explicite
conversions. This makes it hard to actually find those changes and generate TypeInfo
records for the new types aswell, leaving the original records unused.

I've made calls to getTypeInfo(NULL) that would trigger the creation of new TypeInfo records an ICE, but we might want to change the ICE to a warning to avoid breaking code that does not rely on proper TypeInfo.

andralex commented:
@rainers status?

rainers commented:
> @rainers status?

This PR is clearly out of date, but tests with https://github.com/dlang/dmd/pull/6561 shows that the problem still exists.

ibuclaw commented:
@rainers - planning on rebasing?

rainers committed:
do not assume zero-init of structs, calculate lazily if necessary

rainers committed:
fix issue 10442: predict type info references in glue layer

rainers committed:
TypeInfo generation always needs semantic3 to evaluate RTInfo
allow deferred generation of RTInfo

rainers committed:
white space fixes

rainers committed:
do not write type info for broken definitions

rainers committed:
cannot skip RTInfo generation for deprecated structs, they might still be used in another build

rainers commented:
> @rainers - planning on rebasing?

Reimplemented by porting most changes to current master.

One thing I noticed along the way: RTInfo is disabled for deprecated structs if deprecation are considered errors (e.g. when phobos is built). This causes bad RTInfo if these structs are later used with deprecations allowed. I'd recommend not disallowing deprecations when building a library.

rainers committed:
fix rebase

ibuclaw commented on src/dmd/todt.d:
```diff
@@ -1389,6 +1389,9 @@ private extern (C++) class TypeInfoDtVisitor : Visitor
 
         // xgetRTInfo
+        if(!sd.getRTInfo)
+            error(sd.loc, "ICE: RTInfo not evaluated for %s", sd.toChars());
```
It's either an ICE or an error.  Pick one for the sake of coverage. :-)

rainers replied:
It is an ICE, but I was thinking of making it a warning instead, because it is likely to show up on user code that doesn't care about RTInfo (yet).

ibuclaw commented:
Needs tests, assumedly from both issues referenced (however, dlang-bot hasn't picked up the fact that you are fixing them, so some commit messages may be wrong).

rainers committed:
add test case for issue 10442

wilzbach commented:
FYI the bot has a rather simplistic logic - fetch the first page of comments:
- If there's a comment if itself -> UPDATE
- otherwise POST

The easiest workaround would be to open a new PR - I think this is one of the last PRs which are old enough to have been submitted before the bot's existence.

JinShil commented:
I really want to review this PR, but what is RTInfo?  Here's what the docs say:  https://dlang.org/phobos/object.html#.RTInfo

rainers commented:
Rebased and added test. As issue 10786 seem to have been fixed elsewhere, I removed it from the title.

This PR probably contains more than actually needed to fix the specified bug, but other issues only show with a more complex definition of RTInfo(T).

rainers commented:
> I really want to review this PR, but what is RTInfo?

RTInfo is library defined information added to the type info. Its original purpose is to provide information for precise GC scanning (as used in https://github.com/dlang/druntime/pull/1977/files#diff-a68e58fcf0de5aa198fcaceafe4e8cf9R3394), but it has also been used for different purposes.

JinShil commented on src/dmd/aggregate.h:
```diff
@@ -161,7 +162,7 @@ struct StructFlags
 {
 public:
-    int zeroInit;               // !=0 if initialize with 0 fill
+    int zeroInit;               // -1 if still unknown, 1 if initialize with 0 fill, 0 otherwise
```
Can we please start using enums for things like this?

rainers replied:
I didn't yet do it to reduce the diff, but now added an enum.

JinShil commented on src/dmd/toobj.d:
```diff
@@ -521,6 +521,9 @@ void toObjFile(Dsymbol ds, bool multiobj)
 
             // m_RTInfo
+            if(!cd.getRTInfo)
+                error(cd.loc, "ICE: RTInfo not evaluated for %s", cd.toChars());
```
I don't care for the use of "ICE" in the error message; it's too compiler-jargony and seems redundant.  Why isn't a generic error sufficient.  i.e. why not simply print "\`RTInfo\` was not evaluated for \`%s\`"?  Or, maybe this should be an `assert` instead?

P.S.  Don't forget the tick marks.

rainers replied:
I've removed the ICE errors as it goes against the recent trend: if it's not defined in the library, don't emit it.

JinShil commented on src/dmd/typinf.d:
```diff
@@ -53,7 +53,9 @@ extern (C++) void genTypeInfo(Type torig, Scope* sc)
          */
-        if (!builtinTypeInfo(t))
+        if (t.ty == Terror)
+            t.vtinfo.errors = 1;
```
Magic number "1"; what does that mean?  Is it a count of 1 or a integer masquerading as a boolean?  But, if this is the status quo in DMD, I suspect we can change that in another PR.

rainers replied:
`errors` has been converted to bool since, so the magic means "true". Updated.

rainers committed:
use enum for zeroInit
remove error messages for non-existing RTInfo

JinShil reviewed (changes requested):
@rainers I'm sorry about the lack of substance in my review.  I'm still trying to digest all the history, comments, and purpose behind this `RTInfo` feature.  I'll be visiting this PR repeatedly over the next several days, and hopefully reach enough understanding to make a determination.  Of course if someone else wants to make the determination, I certainly won't mind.

jacob-carlborg commented on src/dmd/dstruct.d:
```diff
@@ -221,12 +219,19 @@ enum StructPOD : int
     fwd,   // POD not yet computed
 }
 
+enum ZeroInit : int
```
Does it need to be public?

rainers replied:
Good idea, I'll make it private.

rainers commented on src/dmd/astbase.d:
```diff
@@ -260,6 +260,13 @@ struct ASTBase
         fwd,   // POD not yet computed
     }
 
+    enum ZeroInit : int
```
astbase.d seems stripped of all comments, so probably not good to add one for ZeroInit. I'm not sure why the zeroInit member is kept in these classes to begin with.

rainers commented on src/dmd/dstruct.d:
```diff
@@ -570,6 +561,34 @@ extern (C++) class StructDeclaration : AggregateDeclaration
         return (ispod == StructPOD.yes);
     }
 
+    final bool isZeroInit()
```
ok, will add.

rainers commented on src/dmd/dstruct.d:
```diff
@@ -570,6 +561,34 @@ extern (C++) class StructDeclaration : AggregateDeclaration
+        return zeroInit == ZeroInit.yes;
+    }
+
+    final ZeroInit calcZeroInit()
```
Can be private, too. It cannot be const as it calls isZeroInit on the fields.

rainers commented on src/dmd/expression.d:
```diff
@@ -3699,6 +3708,15 @@ extern (C++) final class AssocArrayLiteralExp : Expression
     }
 }
 
+void verifyTypeInfo(AssocArrayLiteralExp aae, Scope* sc)
```
I'll add it. ditto doesn't work because the functions are too far apart.

rainers committed:
make some symbols private, add ddoc comments

rainers commented:
> I'm sorry about the lack of substance in my review.

No problem. This PR is so old, that I also don't recall all the complications. I didn't want to let the information go to waste, though.

Not sure whether the zeroInit changes are vital to fix RTInfo anymore, but it needs fixing anyway.

The important change here: any time the glue layer (e2ir.d) calls genTypeInfo, the respective type info needs to already exist, because no semantic analysis can be done anymore to evaluate RTInfo(T). That's why the semantic analysis needs to pessimistically predict what the glue layer will do. It's a pretty broken design, but fixing it (moving all glue layer lowering into the semantic phase) will be a disrupting change.

JinShil commented on src/dmd/expression.d:
```diff
@@ -3635,6 +3635,23 @@ extern (C++) final class ArrayLiteralExp : Expression
+ *        sc = the scope in which the type info should be analyzed. If null,
+ *             the global scope of Module.rootModule is used
+ */
+void verifyTypeInfo(ArrayLiteralExp ae, Scope* sc)
```
Is there any particular reason why these `verifyTypeInfo` functions are module-level functions?  Would it be more appropriate for them to be member functions of `ArrayLiteralExp`, `AssocArrayLiteralExp`, etc...?

rainers replied:
Not a particular reason, I've just been trying to follow the trend to keep members in AST classes to a minimum.
I just notice that verifyTypeInfo is never called for CatEp and EqualExp, that must have been gone during the rebase (maybe because semanticTypeInfo is now already being called). I'll try to figure out whether they are obsolete now...

JinShil commented:
> I'll try to figure out whether they are obsolete now

Just waiting to hear back on that.

RazvanN7 commented:
@rainers What is the status of this? Let's bring this to the finish line :dagger:

AndrewEdwards commented:
@rainers Let's move this along the pipeline. This review is awaiting you input regarding the module vs function level appropriateness of verifyTypeInfo. Please resolve an all conflicts and rebase if necessary in preparation for merger. Once this is complete, I will reach out to Walter and David for input.

rainers committed:
fix rebase

rainers commented:
rebased, but the RTInfo generation seems to be omitted even more by now.

AndrewEdwards commented:
@rainers, thanks for the swift response. Some work remains, though, since all 35 checks failed.

AndrewEdwards commented:
Ping @rainers

rainers committed:
no error if RTInfo not defined
resurrect RTInfo generation if TypeInfoStructDeclaration from imported struct added to module
fix TypeInfoStructDeclaration if referred to with modifier

rainers committed:
RTInfo generation must not be omitted for deprecated aggregates

rainers committed:
fix semantic not run on TypeInfo when added to module that has already been processed

rainers committed:
cannot avoid running semantic on struct declaration if template RTInfo is declared

rainers committed:
adapt test to new RTInfo, skip semanticTypeInfo with -betterC, ensure xops built for TypeInfo_Struct

rainers committed:
don't generate TypeInfo data for forward declared struct

wilzbach commented:
dlang-bot keeps commenting, because it fails to find its comment in the first 100 comments and thus thinks it must add the initial comment.
The only way to fix this is opening this up as a new PR, sorry.

rainers committed:
remove unnecessary call to verifyTypeInfo

rainers commented:
> dlang-bot keeps commenting, because it fails to find its comment in the first 100 comments and thus thinks it must add the initial comment.

Thanks for the info, I was already wondering. But it didn't bother me too much ;-)

rainers committed:
static foreach: no need to convert static array, can be used directly as an array

rainers committed:
use semanticTypeInfo instead of direct call to genTypeInfo

AndrewEdwards commented:
@dnadlinger, @WalterBright, @kinke, @Geod24 , @RazvanN7

Gentlemen, this request is pending your review/approval

rainers commented:
While this passes now (but for some styling and C++ header updates), let me point out a couple of thoughts:

- I can extract a couple of fixes into separate PRs
- this shows some of the patches that will be necessary for generic TypeInfo definition as a template under the current design.
- it might generate some more TypeInfo's pessimistically that might not always be necessary, but that's hard to predict. So it might have some (probably small) impact on compilation time
- additional TypeInfo's can introduce other dependencies that might cause link errors that were not there before (but mostly should have been)
- I expect that erroring out if the RTInfo is not generated will happen in the wild, it might be better to make it a warning/message only to avoid breaking code that doesn't care about correct TypeInfo.

I'm not a fan of the current design, though: eager generation of TypeInfo trying to predict what the glue layer might do is error-prone and inefficient.
I imagine that the lowerings to D runtime calls done in the glue layer (and most of the lowerings done in the semantic-passes) should be done in some semantic4-pass ofter semantic3 is complete:
- if "user-code" is error-free after the semantic3-pass, correct lowerings should not produce other (confusing) errors
- semantic4 can still call other semantics on the new code, so e.g. TypeInfo is only generated if the actual call is generated.
- semantic4 could also be the place for templated associative and dynamic array implementations
- CTFE still needs to run in the semantic3-pass
- it will reduce the code of the glue layer that needs to be reimplemented in LDC and GDC.

AndrewEdwards commented:
@dnadlinger, @WalterBright, @kinke, @Geod24 , @RazvanN7

Gentlemen, this request needs your attention, review and approval

Imperatorn commented:
Omg, is this from 2013? 😱

MoonlightSentinel commented on src/dmd/dstruct.d:
```diff
@@ -585,6 +567,52 @@ extern (C++) class StructDeclaration : AggregateDeclaration
+     * Returns:
+     *     true if struct is all zeroes
+     */
+    final bool isZeroInit()
```
Why add this function alongside `TypeStruct.isZeroInit`?

MoonlightSentinel commented on src/dmd/aggregate.h:
```diff
@@ -57,6 +57,13 @@ enum Abstract
 };
 
 FuncDeclaration *search_toString(StructDeclaration *sd);
+enum ZeroInit
```
`enum class` can now be used

MoonlightSentinel commented on src/dmd/semantic3.d:
```diff
@@ -1441,6 +1441,34 @@ private extern(C++) final class Semantic3Visitor : Visitor
         }
     }
 
+    final void generateRTInfo(AggregateDeclaration ad, Scope* sc)
```
DDoc comment

MoonlightSentinel reviewed (commented):
Some general notes/questions:

- why does this PR add more eager `TypeInfo` generation (e.g. `semanticTypeInfo` in `dcast.d`) as well as more lazy generation (e.g. `verifyTypeInfo` in `expressionsem.d`).
- refactorings should be done as separate PR's
- this needs a lot more tests coverage (see CodeCov)

12345swordy commented:
cc @AndrewEdwards @RazvanN7

Imperatorn commented:
@rainers What needs fixing before we can merge? ☀️

Imperatorn commented:
A decade has passed

rainers commented:
@schveiguy I tried the test case from this PR a couple of days ago, and it still failed. I haven't investigated further but I suspect most of the changes here are obsolete as less typeinfo is required now, so semanticTypeInfo() calls could even be removed. Even if I can figure this out, a new PR will probably simpler...

schveiguy commented:
Alright, let's close it to get it off the todo list.
